### PR TITLE
[Updated] Feature/quest system

### PR DESCRIPTION
**Pull request by sboigelot**
Inspired by @quill18 LP of Prison Architect: a quest system for ProjectPorcupine

Choose quest in a list of xml defined quest with LUA goals and reward
image

Keep an eye on your quest status at all time and complete them step by step;
image

Fully expandable by the community

> _I just wanted to work on something else then trade today :p_

### CODE REVIEW

**Comment by sboigelot** [comment]
Note: the changes in the _SCENE_ is only about adding the Quest button on the bottom, easy to re-do if needed

**Comment by vogonistic** [comment]
Can you create a Prefab of it and load that from code instead? We need to get away from changing the scene, it breaks multiple PR's every day.

File: Assets/Scripts/Models/GameEventActions.cs [hunk 1]
```diff
@@ -13,6 +13,7 @@
 using MoonSharp.RemoteDebugger.Network;
 using UnityEngine;
 
+
```
Comment (NogginBops) [inline Assets/Scripts/Models/GameEventActions.cs]: This newline should not be here... Otherwise very cool addition!

**Comment by sboigelot** [comment]
There is no UI from code for the bottom menu, I did that for the dialog box though.
I don't think refactoring the bottom build menu should be part of this PR. It should be a separate one.

I can't just build a prefab for it as I fill a new property of the Menu Controller with the instance of the button. (It's just a simple UI Button)

image

But tell me if this is really needed for this PR according to you and I'll add a dirty button creation in the Menu Controller for now. (@vogonistic)

File: Assets/Scripts/Models/GameEventActions.cs (hunk 1 above)
Comment (sboigelot) [inline Assets/Scripts/Models/GameEventActions.cs]: Thanks ^^

I removed it in https://github.com/TeamPorcupine/ProjectPorcupine/pull/767/commits/5bdca856d5a3688639875984289f1fb2b46a24e8

**Comment by vogonistic** [comment]
That makes sense. And I agree that you shouldn't do that work from this PR. 

Really cool!

**Comment by sboigelot** [comment]
Thanks :+1: 

I can look at improving the UI anyway later. But I will first list all the PR and contributes that are currently working on the same issue. I feel I see every day a PR or two about that topic.

**Comment by ComputerOverlord** [comment]
It's good, but when I get my KR system up and running, It will be repurposed.

**Comment by sboigelot** [comment]
@ComputerOverlord No problem, fell free to improve/extend/replace it :) We are all working together for the greater good ^^

**Comment by dustinwloring1988** [comment]
Love the quest idea.
As to adding the button though code instead of chaning the UI.

- Make a prefab of the button
- Save it in the resource folder
- Now in the MenuController script instantiate the button and make the bottom menu the parent.

**Comment by bjubes** [comment]
@dustinwloring1988  is right, you don't have to modify the scene when adding the prefab because the Menu Controller fills its inspector fields at run time, not before you press play. Just add a public field for it and leave it blank in the inspector, and instantiate it programmatically.

**Comment by bjubes** [comment]
also @dustinwloring1988  if you could make an entry about how to add ui to the scene in the[ "how does it work?" section on the wiki](https://github.com/TeamPorcupine/ProjectPorcupine/wiki/How-does-it-work%3F) that would help out tons in the future

**Comment by dustinwloring1988** [comment]
@bjubes   No Problem will do.

**Comment by quill18** [comment]
One of the great things about "Quests" like this, as in Prison Architect, is that they actually serve as a tutorial/guide for the first part of the game, without being boring/limiting like a real tutorial. Excellent idea.

**Comment by sboigelot** [comment]
Done, the _SCENE_ is reverted to the upstream/master one and I added a method to create buttons in the MenuController

**Comment by sboigelot** [comment]
@vogonistic I used a copy of the master scene, but somehow git still believe there is changes. If you merge this and you have conflict, feel free to use the master scene to override the one from the feature branch. (but "This branch has no conflicts with the base branch" for now)

**Comment by dustinwloring1988** [comment]
Cant wait until this get added.

**Comment by koosemose** [comment]
Definitely like the idea of a quest system, especially for the tutorial aspect as quill mentioned, but not a fan of the always on aspect, it eats way too much screen space. Would it be possible to make it not always on and either allow it to be set to always show or just a "pin" system where you can pin the particular quest you want to keep track of to stay up?

**Comment by ComputerOverlord** [comment]
@koosemoose It wouldn't eat up screen space--does PA's always-on system eat up screen space?

**Comment by ComputerOverlord** [comment]
Is there a OnRoomBuilt function/method you could use for a first quest

**Comment by vogonistic** [comment]
Please fix the merge conflicts and remove `[conflict]` from the title when it's done. Thanks!

**Comment by sboigelot** [comment]
@koosemose I just added a toggle to show/hide the quest on the main screen

image

image

@vogonistic conflict resolved

**Comment by koosemose** [comment]
Awesome, thank you very much. Though you do have a typo (or it's a quest for pigs ;D), SowQuests.

**Comment by sboigelot** [comment]
@koosemose fixed

**Comment by koosemose** [comment]
Of course, now if we ever get animals for any reason, I really want a Sow Quest :D

**Comment by quill18** [comment]
Merging!

**Comment by sboigelot** [comment]
:tada:  Thanks

**Pull request merged**